www/js/plugin/storage.ts

```diff
@@ -0,0 +1,182 @@
+        localStorageSet(key, uc_stored_val);
+        return uc_stored_val;
+      } else if (uc_stored_val == null) {
+        console.assert(ls_stored_val != null);
+        /*
+         * Backwards compatibility ONLY. Right after the first
+         * update to this version, we may have a local value that
+         * is not a JSON object. In that case, we want to munge it
+         * before storage. Remove this after a few releases.
+         */
+        ls_stored_val = mungeValue(key, ls_stored_val);
+        displayErrorMsg(`Local ${key} found, native ${key} missing, writing ${key} to native`);
+        logDebug(`for key ${key}, uc_stored_val = ${JSON.stringify(uc_stored_val)},
+                                      ls_stored_val = ${JSON.stringify(ls_stored_val)}.
+                                      copying local ${key} to native...`);
+        return window['cordova'].plugins.BEMUserCache.putLocalStorage(key, ls_stored_val).then(() => {
+          // we only return the value after we have finished writing
+          return ls_stored_val;
+        });
+      }
+      console.assert(ls_stored_val != null && uc_stored_val != null,
+        "ls_stored_val =" + JSON.stringify(ls_stored_val) +
+        "uc_stored_val =" + JSON.stringify(uc_stored_val));
+      displayErrorMsg(`Local ${key} found, native ${key} found, but different,
+                           writing ${key} to local`);
+      logDebug(`for key ${key}, uc_stored_val = ${JSON.stringify(uc_stored_val)},
+                    ls_stored_val = ${JSON.stringify(ls_stored_val)}.
+                    copying native ${key} to local...`);
+      localStorageSet(key, uc_stored_val);
+      return uc_stored_val;
+    }
+  });
+}
+
+export function storageSet(key: string, value: any) {
+  const storeVal = mungeValue(key, value);
+  /*
+   * How should we deal with consistency here? Have the threads be
+   * independent so that there is greater chance that one will succeed,
+   * or the local only succeed if native succeeds. I think parallel is
+   * better for greater robustness.
+   */
+  localStorageSet(key, storeVal);
+  return window['cordova'].plugins.BEMUserCache.putLocalStorage(key, storeVal);
+}
+
+export function storageGet(key: string) {
+  return getUnifiedValue(key).then((retData) => unmungeValue(key, retData));
+}
+
+export function storageRemove(key: string) {
+  localStorage.removeItem(key);
+  return window['cordova'].plugins.BEMUserCache.removeLocalStorage(key);
+}
+
+export function storageClear({ local, native }: { local?: boolean, native?: boolean }) {
+  if (local) localStorage.clear();
+  if (native) return window['cordova'].plugins.BEMUserCache.clearAll();
+  return Promise.resolve();
+}
+
+export function storageGetDirect(key: string) {
+  // will run in background, we won't wait for the results
+  getUnifiedValue(key);
+  return unmungeValue(key, localStorageGet(key));
+}
+
+function findMissing(fromKeys, toKeys) {
```

Comment:
It would be good to add a unit test for `findMissing` as well - it is covered in `storageSyncLocalAndNative`, but sometimes you can test more extensively by calling the lower function directly (e.g. call it with a blank list, with null etc)